docker/models/containers.py

```diff
@@ -709,7 +716,14 @@ def run(self, image, command=None, stdout=True, stderr=False,
         if exit_status != 0:
             stdout = False
             stderr = True
-        out = container.logs(stdout=stdout, stderr=stderr)
+
+        logging_driver = container.attrs['HostConfig']['LogConfig']['Type']
+
+        if logging_driver == 'json-file' or logging_driver == 'journald':
```

Comment:
@thaJeztah It always comes down to the same issue - the API direly needs error codes if we want to be able to handle these things gracefully. Until then, I'm afraid hardcoded values are the best solution here. Especially in this case, `run`'s purpose is to be a convenience method that imitates `docker run` - for more fine-grained control, people will still be able to call `logs` on the container directly and handle the possible exception themselves.

As for `logs` over `attach`, the reason is 2-fold:
1. We don't have an interactive mode, so both are functionally equivalent
2. Historically `logs` works better than `attach`